### PR TITLE
feat: new data source cloudavenue bms

### PR DESCRIPTION
**Pull request by dmicheneau**
### Description of your changes

If you submit change in the provider code, please make sure to:

- [x] Write or modify examples in `examples/` directory
- [x] Write or modify acceptance tests
- [x] Run `make generate` to ensure the doc was updated properly

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -24,7 +24,7 @@ require (
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.34.0
 	github.com/iancoleman/strcase v0.3.0
 	github.com/kr/pretty v0.3.1
-	github.com/orange-cloudavenue/cloudavenue-sdk-go v0.11.1
+	github.com/orange-cloudavenue/cloudavenue-sdk-go v0.12.0
```
Comment (azrod) [inline go.mod:27]: Add changelog for this

File: internal/provider/bms/bms_schema.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+			MarkdownDescription: "The `cloudavenue_bms` data source allows you to retrieve information about your Bare Metal Server.",
+		},
+		Attributes: map[string]superschema.Attribute{
+			"timeouts": superschema.TimeoutAttribute{
+				// Resource: &superschema.ResourceTimeoutAttribute{
+				// 	Create: true,
+				// 	Update: true,
+				// 	Delete: true,
+				// 	Read:   true,
+				// },
```
Comment (azrod) [inline internal/provider/bms/bms_schema.go:23]:
```suggestion
```

File: terraform-provider-cloudavenue
Comment (azrod) [inline terraform-provider-cloudavenue:1]: Remove

File: go.mod (hunk 1 above)
Comment (dmicheneau) [reply go.mod:27]: fix

File: internal/provider/bms/bms_schema.go (hunk 2 above)
Comment (dmicheneau) [reply internal/provider/bms/bms_schema.go]: fix

File: internal/provider/bms/bms_datasource.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	defer cancel()
+
+	// Read data from the API
+	bms, err := d.client.CAVSDK.V1.BMS.List()
+	if err != nil {
+		resp.Diagnostics.AddError("error on list BMS(s)", err.Error())
+		return
+	}
+	if len(*bms) == 0 {
+		config.Env.Set(ctx, []*bmsModelDatasourceEnv{})
```
Comment (azrod) [inline internal/provider/bms/bms_datasource.go:96]: This is unnecessary the value is already `null`. The null value is define in `Config.Get()` (Line 65).

File: internal/provider/bms/bms_types.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	}
+}
+
+// SetID sets the ID.
+func (m *bmsModelDatasource) SetID(id *string) {
+	m.ID.Set(*id)
+}
+
+// SetNetwork sets the Network of BMS listed.
+func (m *bmsModelDatasource) SetNetwork(ctx context.Context, bms *v1.BMS) (net []*bmsModelDatasourceNetwork, err error) {
```
Comment (azrod) [inline internal/provider/bms/bms_types.go:68]: The function name doesn't seem appropriate to me. Indeed we do not "Set" values ​​in the struct `bmsModelDatasource` but we construct a terraform compatible object.  

I propose to rename it to `NetworkToTerraform` and this function is not method of `bmsModelDatasource`. 

The `ctx` is unused. 

```suggestion
func NetworkToTerraform(bms *v1.BMS) (net []*bmsModelDatasourceNetwork, err error) {
```


Suggest simplified func 
```go 
// NetworkToTerraform converts the network information to Terraform format.
func NetworkToTerraform(bms *v1.BMS) (net []*bmsModelDatasourceNetwork) {

	for _, network := range bms.GetNetworks() {
		var x bmsModelDatasourceNetwork

		x.VLANID.Set(network.VLANID)
		x.Subnet.Set(network.Subnet)
		x.Prefix.Set(network.Prefix)
		net = append(net, &x)
	}

	return
}
```

File: internal/provider/bms/bms_types.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	networks := bms.GetNetworks()
+	if len(networks) == 0 {
+		return make([]*bmsModelDatasourceNetwork, 0), nil
+	}
```
Comment (azrod) [inline internal/provider/bms/bms_types.go:72]: This is unnessarry. if networks have not entry the for loop is ignored.

File: internal/provider/bms/bms_types.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+		x.Subnet.Set(network.Subnet)
+		x.Prefix.Set(network.Prefix)
+		net = append(net, &x)
+	}
+
+	return net, nil
+}
+
+// SetBMS sets the BMS information.
+func (m *bmsModelDatasource) SetBMS(ctx context.Context, bms *v1.BMS) (bmsList []*bmsModelDatasourceBMS, err error) {
```
Comment (azrod) [inline internal/provider/bms/bms_types.go:85]: Same `SetNetwork`

File: internal/provider/bms/bms_types.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+		OS                supertypes.StringValue                                             `tfsdk:"os"`
+		BiosConfiguration supertypes.StringValue                                             `tfsdk:"bios_configuration"`
+		Storage           supertypes.SingleNestedObjectValueOf[bmsModelDatasourceBMSStorage] `tfsdk:"storage"`
+	}
+
+	bmsModelDatasourceBMSStorage struct {
+		Local  supertypes.ListNestedObjectValueOf[bmsModelDatasourceBMSStorageGen] `tfsdk:"local"`
+		System supertypes.ListNestedObjectValueOf[bmsModelDatasourceBMSStorageGen] `tfsdk:"system"`
+		Data   supertypes.ListNestedObjectValueOf[bmsModelDatasourceBMSStorageGen] `tfsdk:"data"`
+		Shared supertypes.ListNestedObjectValueOf[bmsModelDatasourceBMSStorageGen] `tfsdk:"shared"`
```
Comment (azrod) [inline internal/provider/bms/bms_types.go:44]: Why not use `SetNestedObjectValueOf` ?

File: internal/provider/bms/bms_types.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	bmsModelDatasourceBMSStorageGen struct {
```
Comment (azrod) [inline internal/provider/bms/bms_types.go:47]:
```suggestion
	bmsModelDatasourceBMSStorageDetail struct {
```

File: internal/provider/bms/bms_types.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+		ID:       supertypes.NewStringNull(),
+		Timeouts: timeoutsD.Value{},
+		Env:      supertypes.NewSetNestedObjectValueOfNull[bmsModelDatasourceEnv](ctx),
+	}
+}
+
+// SetID sets the ID.
+func (m *bmsModelDatasource) SetID(id *string) {
+	m.ID.Set(*id)
+}
```
Comment (azrod) [inline internal/provider/bms/bms_types.go:65]: This is really necessary ?

**Review by azrod — CHANGES_REQUESTED** [review]

File: internal/provider/bms/bms_types.go (hunk 9 above)
Comment (dmicheneau) [reply internal/provider/bms/bms_types.go]: No, the line was deleted

File: docs/data-sources/bms.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+- `bms` (Attributes Set) The BMS list. (see [below for nested schema](#nestedatt--env--bms))
+- `network` (Attributes Set) The network array for all BMS listed. (see [below for nested schema](#nestedatt--env--network))
+
+<a id="nestedatt--env--bms"></a>
+### Nested Schema for `env.bms`
+
+Read-Only:
+
+- `bios_configuration` (String) The BIOS configuration of the BMS.
+- `bms_type` (String) The type of the BMS.
```
[the remark this answers is not in the document]
Comment (dmicheneau) [reply docs/data-sources/bms.md]: Fix

File: internal/provider/bms/bms_datasource.go (hunk 3 above)
Comment (dmicheneau) [reply internal/provider/bms/bms_datasource.go]: Fix

File: internal/provider/bms/bms_types.go (hunk 4 above)
Comment (dmicheneau) [reply internal/provider/bms/bms_types.go]: Fix

File: internal/provider/bms/bms_types.go (hunk 5 above)
Comment (dmicheneau) [reply internal/provider/bms/bms_types.go]: Fix

File: internal/provider/bms/bms_types.go (hunk 6 above)
Comment (dmicheneau) [reply internal/provider/bms/bms_types.go]: Fix

File: internal/provider/bms/bms_types.go (hunk 7 above)
Comment (dmicheneau) [reply internal/provider/bms/bms_types.go]: Fix

File: internal/provider/bms/bms_types.go (hunk 8 above)
Comment (dmicheneau) [reply internal/provider/bms/bms_types.go]: Gen for generic, but detail is the good name use in sdk too: Fix

File: internal/provider/bms/bms_types.go [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+		x.Subnet.Set(network.Subnet)
+		x.Prefix.Set(network.Prefix)
+		net = append(net, &x)
+	}
+
+	return
+}
+
+// Put API BMS information to Terraform Object.
+func BMSToTerraform(ctx context.Context, bms *v1.BMS) (bmsList []*bmsModelDatasourceBMS) { //nolint: revive
```
Comment (azrod) [inline internal/provider/bms/bms_types.go:76]:
```suggestion
func BMSToTerraform(bms *v1.BMS) (bmsList []*bmsModelDatasourceBMS) {
```

File: internal/provider/bms/bms_types.go [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+	storage.System.Set(ctx, setStorageDetail(storages.GetSystem()))
+	storage.Data.Set(ctx, setStorageDetail(storages.GetData()))
+	storage.Shared.Set(ctx, setStorageDetail(storages.GetShared()))
+	return
+}
+
+func setStorageDetail(storage []v1.BMSStorageDetail) (storageDetail []*bmsModelDatasourceBMSStorageDetail) {
+	if len(storage) == 0 {
+		return
+	}
```
Comment (azrod) [inline internal/provider/bms/bms_types.go:102]:
```suggestion
```

Comment (dmicheneau) [reply internal/provider/bms/bms_types.go:102]: Fix

**Review by azrod — APPROVED** [review]

**Pull request merged**